src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -73751,6 +73751,223 @@ static void M2(Enumerable e)
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "e").WithLocation(13, 29));
         }
 
+        [Fact]
+        [WorkItem(38638, "https://github.com/dotnet/roslyn/issues/38638")]
+        public void ForEach_ExtensionGetEnumerator1()
+        {
+            var source = @"
+using System.Collections.Generic;
+class A<T>
+{
+    public static void M()
+    {
+        foreach(var s in new A<string?>())
+        {
+            _ = s.ToString();
+        }
+    }
+}
+static class Extensions
+{
+    public static IEnumerator<T> GetEnumerator<T>(this A<T> a) => throw null!;
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue(), parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (9,17): warning CS8602: Dereference of a possibly null reference.
+                //             _ = s.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(9, 17));
+        }
+
+        [Fact]
+        [WorkItem(38638, "https://github.com/dotnet/roslyn/issues/38638")]
+        public void ForEach_ExtensionGetEnumerator2()
+        {
+            var source = @"
+using System.Collections.Generic;
+class A<T>
+{
+    public static void M()
+    {
+        foreach(var s in new A<string>())
+        {
+            _ = s.ToString();
+        }
+    }
+}
+static class Extensions
+{
+    public static IEnumerator<T> GetEnumerator<T>(this A<T> a) => throw null!;
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue(), parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(38638, "https://github.com/dotnet/roslyn/issues/38638")]
+        public void ForEach_ExtensionGetEnumerator3()
+        {
+            var source = @"
+using System.Collections.Generic;
+class A<T>
+{
+    public static void M()
+    {
+        foreach(var s in new A<string>())
+        {
+            _ = s.ToString();
+        }
+    }
+}
+static class Extensions
+{
+    public static IEnumerator<T?> GetEnumerator<T>(this A<T> a) where T : class => throw null!;
```

Comment:
Consider adding a variation with the `notnull` constraint. Probably would work to just make this a Theory and parameterize.

Reply:
I'm not sure what you mean by "a variation with the `notnull` constraint". Is that different to ForEach_ExtensionGetEnumerator6?

Reply:
Basically a version where `T` is constrained to `notnull`.